tests/robots/test_robot.py

```diff
@@ -289,6 +290,51 @@ def test_get_configurable_joints_wo_semantics(panda_robot_instance_wo_semantics)
     assert all(matches)
 
 
+def test_semantics_serialization(panda_srdf, panda_urdf):
+    model = RobotModel.from_urdf_file(panda_urdf)
+    semantics = RobotSemantics.from_srdf_file(panda_srdf, model)
+    semantics_string = json_dumps(semantics)
+    semantics2 = json_loads(semantics_string)
+    assert isinstance(semantics, RobotSemantics)
+    assert isinstance(semantics2, RobotSemantics)
+    for disabled_collision in semantics.disabled_collisions:
+        assert disabled_collision in semantics2.disabled_collisions
+
+
+def test_robot_serialization(panda_robot_instance):
+    robot = panda_robot_instance
+    robot.scale(0.001)
+    robot_string = json_dumps(robot)
+    robot2 = json_loads(robot_string)
+    robot2_string = json_dumps(robot2)
+    assert robot2_string != ""
+    assert isinstance(robot, Robot)
+    assert isinstance(robot2, Robot)
+    assert robot.model.name == robot2.model.name
+    assert robot._scale_factor == robot2._scale_factor
+    for index, joint in enumerate(robot.model.joints):
+        assert joint.name == robot2.model.joints[index].name
+        assert str(joint.origin) == str(robot2.model.joints[index].origin)
+    for index, link in enumerate(robot.model.links):
+        assert link.name == robot2.model.links[index].name
+
+
+def test_robot_serialization_with_tool(ur5_robot_instance, robot_tool1):
+    robot = ur5_robot_instance
+    tool = robot_tool1
+    robot.attach_tool(tool)
+    robot_string = json_dumps(robot)
+    robot2 = json_loads(robot_string)
+    for tool in robot2.attached_tools.values():
+        assert isinstance(tool, Tool)
+    assert len(robot2.attached_tools) == 1
+    assert robot.main_group_name == robot2.main_group_name
+    tool1 = robot.attached_tool
+    tool2 = robot2.attached_tool
+    assert tool1.name == tool2.name
+    assert str(tool1.frame) == str(tool2.frame)
+
+
```

Comment:
I'd add another unittest checking deepcopy. should behave the same as (de)serializing but sometimes doesn't

Reply:
I tried deepcopy but It doesn't seem to work. I think some years ago, I heard from someone to not use deepcopy on these objects. I don't remember why. But basically I never deepcopied any compas objects since.

Just to give you a taste of that error message:

```
./tests/robots/test_robot.py::test_robot_deepcopy_with_tool Failed: [undefined]AttributeError: 'Visual' object has no attribute '_guid'
tests\robots\test_robot.py:345: in test_robot_deepcopy_with_tool
    robot2 = deepcopy(robot)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:206: in _deepcopy_list
    append(deepcopy(a, memo))
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:271: in _reconstruct
    state = deepcopy(state, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:231: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:146: in deepcopy
    y = copier(x, memo)
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:206: in _deepcopy_list
    append(deepcopy(a, memo))
..\..\..\anaconda3\envs\cf-dev\Lib\copy.py:161: in deepcopy
    rv = reductor(4)
..\..\..\anaconda3\envs\cf-dev\Lib\site-packages\compas\data\data.py:121: in __getstate__
    "guid": str(self.guid),
..\..\..\anaconda3\envs\cf-dev\Lib\site-packages\compas\data\data.py:200: in guid
    if not self._guid:
E   AttributeError: 'Visual' object has no attribute '_guid'
```

Reply:
it's known `deepcopy` errors produce the most beautiful stacks.. 
To this specific one, could there be a class called `Visual` (really?) somewhere which inherits from `Data` but doesn't call `super().__init__()`?

Reply:
https://github.com/compas-dev/compas/blob/ba932fcb55b2839ef493f4181b1f080f377a772b/src/compas/robots/model/link.py#L166-L191

It looks okay here. I'm not sure how this Visual object is created. It seems that there is a Geometry class within it and that is created during `from_data()`:

https://github.com/compas-dev/compas/blob/ba932fcb55b2839ef493f4181b1f080f377a772b/src/compas/robots/model/link.py#L244-L248

Reply:
If someone wants to debug this, this is the test code:

```
def test_robot_deepcopy(panda_robot_instance):
    import copy
    robot = panda_robot_instance
    robot.scale(0.001)
    robot2 = copy.deepcopy(robot)
    assert robot is not robot2

def test_robot_deepcopy2(ur5_robot_instance, robot_tool1):
    import copy
    robot = ur5_robot_instance
    robot.attach_tool(robot_tool1)
    robot2 = copy.deepcopy(robot)
    assert robot is not robot2
```

Both test won't work and they have different error message